### PR TITLE
Intake | Add `established_end_product` method to RampElection

### PR DESCRIPTION
**Pull request by shanear**
This will allow us to see if the end product of a ramp election
is still active or not.

This is just a back end change for now.

Also create `Generators::EndProduct` to make it easier to mock
end products. This change includes scoping mocks to specific
veteran ids to improve test coverage.

connects #3957

### CODE REVIEW

File: app/models/veteran.rb [hunk 1]
```diff
@@ -99,6 +103,10 @@ def set_attrs_from_bgs_record
     end
   end
 
+  def fetch_end_products
```
Comment (shanear) [inline app/models/veteran.rb]: This method also exists almost character for character in `appeal.rb` but makes more sense here in `veteran.rb`.

In the future we could factor them both here.

File: lib/fakes/bgs_service.rb [hunk 2]
```diff
@@ -200,10 +200,13 @@ def self.power_of_attorney_records
...
     self.inaccessible_appeal_vbms_ids = []
+    self.end_product_records = {}
   end
 
-  def get_end_products(_veteran_id)
-    end_product_data || self.class.no_grants
+  def get_end_products(veteran_id)
+    records = self.class.end_product_records || {}
+
+    records[veteran_id] || records[:default] || []
```
Comment (shanear) [inline lib/fakes/bgs_service.rb]: kept the ability to set the EndProducts that return for all veterans (since it's useful for demo)

File: lib/fakes/bgs_service.rb [hunk 3]
```diff
@@ -278,86 +281,4 @@ def default_address
       zip_prefix_nbr: "94103"
     }
   end
-
-  def default_veteran_record
```
Comment (shanear) [inline lib/fakes/bgs_service.rb]: This was an unused method i stumbled on... 😃

File: spec/feature/dispatch/establish_claim_spec.rb [hunk 4]
```diff
@@ -594,22 +592,18 @@
...
 
-      # :nocov:
-      context "When there is an existing 070 EP",
-              skip: "This test hangs somewhat regularly for unknown reasons" do
-        before do
-          BGSService.end_product_data = [
-            {
-              benefit_claim_id: "1",
-              claim_receive_date: 10.days.ago.to_formatted_s(:short_date),
+      context "When there is an existing 070 EP" do
```
Comment (shanear) [inline spec/feature/dispatch/establish_claim_spec.rb]: I'm giving this test another chance, I'll commit to keeping an eye on it. Ran it 20 times with 100% success:

screen shot 2017-12-15 at 5 38 42 pm

File: app/models/end_product.rb [hunk 5]
```diff
@@ -2,6 +2,7 @@ class EndProduct
   include ActiveModel::Model
   include ActiveModel::Validations
 
+  # NOTE: This is not a comprehensive list of possible statuses
```
Comment (nickheiner-usds) [inline app/models/end_product.rb]: :smile:

File: lib/fakes/bgs_service.rb (hunk 3 above)
Comment (nickheiner-usds) [reply lib/fakes/bgs_service.rb]: 🔥

File: spec/feature/dispatch/establish_claim_spec.rb (hunk 4 above)
Comment (nickheiner-usds) [reply spec/feature/dispatch/establish_claim_spec.rb]: Ok, sounds good.

The other thing I like to do is actually push to Travis with `ensure_stable` to see if Travis has the same consistency.

File: spec/feature/switch_user_spec.rb [hunk 6]
```diff
@@ -4,7 +4,7 @@
   before do
     # Switch user only works in demo
     ENV["DEPLOY_ENV"] = "demo"
-    BGSService.end_product_data = BGSService.all_grants
+    BGSService.end_product_records = { default:  BGSService.all_grants }
```
Comment (nickheiner-usds) [inline spec/feature/switch_user_spec.rb]: Out of curiosity, why make this change? Are you expecting to add more keys to this hash eventually?

File: spec/models/ramp_election_spec.rb [hunk 7]
```diff
@@ -123,6 +125,33 @@
     end
   end
 
+  context "#established_end_product" do
+    subject { ramp_election.established_end_product }
+
+    let!(:other_ep) { Generators::EndProduct.build(veteran_file_number: veteran_file_number) }
```
Comment (nickheiner-usds) [inline spec/models/ramp_election_spec.rb]: Why is this necessary?

**Comment by nickheiner-usds** [comment]
The title of this PR is `Intake | Add `establish_end_product` method to RampElection`, but I don't actually see that method being added to `RampElection`. Are we missing something, or is the title inaccurate?

File: spec/models/ramp_election_spec.rb (hunk 7 above)
Comment (shanear) [reply spec/models/ramp_election_spec.rb]: Just to make sure we aren't pulling any ole EP, only the one that matches the `end_product_reference_id`

**Comment by nickheiner-usds** [comment]
The tests LGTM.

File: spec/feature/switch_user_spec.rb (hunk 6 above)
Comment (shanear) [reply spec/feature/switch_user_spec.rb]: Because we look up end products by veteran id in BGS. This allows us to be a little bit more precise with our tests and specify mocked EPs for specific veterans. Formerly we [were mocking EPs that returned for all veterans](https://github.com/department-of-veterans-affairs/caseflow/pull/4177/files#diff-6dd2c923d1bbe7d63927f83644ed12abL206)

File: spec/feature/dispatch/establish_claim_spec.rb (hunk 4 above)
Comment (shanear) [reply spec/feature/dispatch/establish_claim_spec.rb]: Good call, I'm gonna do that in a separate PR because its sort of a separate thing as to not block this

**Comment by shanear** [comment]
> Update Looks like established_end_product is the name of the method you were referring to. Cool, I understand now.

whoopsies. Fixed the title.

File: spec/models/ramp_election_spec.rb (hunk 7 above)
Comment (nickheiner-usds) [reply spec/models/ramp_election_spec.rb]: Ok, cool. Just for my understanding of rspec, how does this get used anywhere if `:other_ep` only appears once in the codebase?

```
 λ git grep other_ep
spec/models/ramp_election_spec.rb:    let!(:other_ep) { Generators::EndProduct.build(veteran_file_number: veteran_file_number) }
```

Comment (shanear) [reply spec/models/ramp_election_spec.rb]: `let` lazily loads the value inside
`let!` will eager load the value inside
